website/www/site/content/en/blog/unit-testing-in-beam.md

```diff
@@ -0,0 +1,197 @@
+      '-120.05,39.37,29.00,4085.00,681.00,1557.00,626.00,6.8085,364700.00'
+    ]
+    with beam.Pipeline() as p2:
+      result = (
+                p2
+                | beam.Create(input_elements)
+                | beam.Map(MapAndCombineTransform())
+        )
+      assert_that(result,equal_to(expected))
+```
+
+###Example 3
+
+Suppose we write a pipeline that reads data from a JSON file, gets passed through a custom function that makes external API calls for parsing, and is then written to a custom destination (for example, we need to do some custom data formatting to have data prepared for a downstream application).
+
+
+The pipeline is structured as follows:
+
+```python
+#The following packages are used to run the example pipelines
+
+import apache_beam as beam
+from apache_beam.io import ReadFromText, WriteToText
+from apache_beam.options.pipeline_options import PipelineOptions
+
+class MyDoFn(beam.DoFn):
+  def process(self,element):
+          returned_record = MyApiCall.get_data("http://my-api-call.com")
+          if len(returned_record)!=10:
+            raise ValueError("Length of record does not match expected length")
+          yield returned_record
+
+with beam.Pipeline() as p3:
+  result = (
+          p3
+          | ReadFromText("/content/sample_data/anscombe.json")
+          | beam.ParDo(MyDoFn())
+          | WriteToText("/content/example1")
+  )
+```
+
+This test checks that if the API response is a record of the wrong length, we throw the expected error.
+
+```python
+!pip install mock  # Install the 'mock' module
+```
+```python
+# We import the mock package for mocking functionality.
+from unittest.mock import Mock,patch
+# from MyApiCall import get_data
+import mock
+
+
+# MyApiCall is a function that calls get_data to fetch some data via an API call.
+@patch('MyApiCall.get_data')
+def test_error_message_wrong_length(self, mock_get_data):
+ response = ['field1','field2']
+ mock_get_data.return_value = Mock()
+ mock_get_data.return_value.json.return_value=response
+
+ input_elements = ['-122.050000,37.370000,27.000000,3885.000000,661.000000,1537.000000,606.000000,6.608500,344700.000000'] #input length 9
+ with self.assertRaisesRegex(ValueError,
+                             "Length of record does not match expected length'"):
+     p3 = beam.Pipeline()
+     result = p3 | beam.create(input_elements) | beam.ParDo(MyDoFn())
+     result
+```
+
+###The following cover other testing best practices:
+
+1) Test all error messages you raise.
+2) Cover any edge cases that might be present in your data.
+3) Notice that in example 1, we could have written the `beam.Map` step with lambda functions:
+
+```
+beam.Map(lambda x: x.strip().split(',')) | beam.Map(lambda x: float(x[8])/float(x[4])
+```
+
+, instead of `beam.Map(median_house_value_per_bedroom)`.  The latter (separating lambdas into a helper function) is the recommended approach for more testable code, as changes to the function would be modularized.
```

Comment:
```suggestion
Separating lambdas into a helper function by using `beam.Map(median_house_value_per_bedroom)` is the recommended approach for more testable code, because changes to the function would be modularized.
```